### ISSUE TITLE
Support column and row-wise outer products for sparse matrices

### ISSUE THREAD

**Issue by thomasahle** [scipy/scipy#10487]
The row-wise outer/tensor product (for csr matrices and column-wise for csc) is a useful way of adding quadratic features to data before using a linear model (such as leas squares with `scipy.sparse.linalg.lsqr`). This is also known as bilinear pooling.

There doesn't currently seem to be a good way to compute these in scipy, even though the output is going to be sparse.
Currently, the only option is to make a non-vectorized loop in python, or convert to dense matrices, perform the operation and then back to sparse.

In many ways, this is similar to the Kronecker product supported by scipy.sparse.
The row- and column-wise products are also related to the [Khatri-Rao product](https://en.wikipedia.org/wiki/Kronecker_product#Related_matrix_operations) and have many useful properties.

**Comment by perimosocordiae** [comment]
Seems like a reasonable feature to include. The implementation in the top answer at https://stackoverflow.com/questions/57099722/row-wise-outer-product-on-sparse-matrices may be a good starting point, and if additional performance is desired it's also possible to write a C++ version.

A few questions to think over:
 1. scikit-learn has some utilities for working with scipy.sparse matrices. Do they have a sparse outer product function already?
 2. where would this new function best fit into the current API?

**Comment by thomasahle** [comment]
1. I believe the scikit-learn functions are the `sklearn.utils.sparsefuncs` below:
```
utils.sparsefuncs.incr_mean_variance_axis(X, …)	Compute incremental mean and variance along an axix on a CSR or CSC matrix.
utils.sparsefuncs.inplace_column_scale(X, scale)	Inplace column scaling of a CSC/CSR matrix.
utils.sparsefuncs.inplace_row_scale(X, scale)	Inplace row scaling of a CSR or CSC matrix.
utils.sparsefuncs.inplace_swap_row(X, m, n)	Swaps two rows of a CSC/CSR matrix in-place.
utils.sparsefuncs.inplace_swap_column(X, m, n)	Swaps two columns of a CSC/CSR matrix in-place.
utils.sparsefuncs.mean_variance_axis(X, axis)
```
None of these appear to overlap with the functionality suggested.

2. To me it would make sense to have the outer product as `scipy.sparse.outer` as a sibling function to `scipy.sparse.kron` and `scipy.sparse.kronsum`. One concern would be whether it should act differently on csr vs csc matrices (outer product on rows in the first case, columns in the second). If that is the case, it might be better to have it live as a method on said matrix types.

Another issue is what to call the product.
Wikipedia mentions it's sometimes called '"face-splitting product" of matrices or the "transposed Khatri-Rao product"', but both are quite a mouthful. Simply calling it "outer product" seems natural, however, this may be a bit confusing next to `numpy.outer` and `scipy.outer`, which both act on individual vectors rather than matrices.

**Comment by thomasahle** [comment]
Should I try to write up a patch?

**Comment by perimosocordiae** [comment]
Sure, getting some code in place will help us (and the greater Scipy community) evaluate whether this is a useful API to expose.

I'd start with a single `scipy.sparse.outer` function that takes an `axis` argument. It can then convert to CSR or CSC as necessary to enable efficient products. Don't worry about optimizing it right away, as this implementation will mostly serve as a reference point for discussion on the mailing list.
